### PR TITLE
Manage contact for registered domains 

### PR DESCRIPTION
**Pull request by DXTimer**
This PR allows for the management of a registered domain's associated contact. To do this we are introducing a new resource called `dnsimple_registered_domain_contact` that leverages the new Registrant Change API to trigger a contact change for the domain.

In addition, there has been added a new data source `dnsimple_registrant_change_check` which allows developers to get the requirements needed in order to carry out the contact change and if it will result in an ownership change at the registry. Refer to the documentation for more information on the data source and the new resource.

Belongs to https://github.com/dnsimple/dnsimple-business/issues/1729
Closes https://github.com/dnsimple/dnsimple-external-integrations/issues/6

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,15 +1,15 @@
 module github.com/terraform-providers/terraform-provider-dnsimple
 
 require (
-	github.com/dnsimple/dnsimple-go v1.2.0
+	github.com/dnsimple/dnsimple-go v1.2.2-0.20230824051107-a0bab607f157
```
Comment (wilson-dnsimple) [inline go.mod:4]: Should this be `v1.3.0`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,13 @@
 
 ## main
 
+## 1.2.0
```
Comment (wilson-dnsimple) [inline CHANGELOG.md:5]:
```suggestion
## 1.2.0 (Unreleased)
```

File: internal/framework/resources/registered_domain/common.go [hunk 3]
```diff
@@ -121,7 +123,17 @@ func (r *RegisteredDomainResource) updateModelFromAPIResponse(ctx context.Contex
 		data.State = types.StringValue(domain.State)
 		data.UnicodeName = types.StringValue(domain.UnicodeName)
 		data.AccountId = types.Int64Value(domain.AccountID)
-		data.ContactId = types.Int64Value(domain.RegistrantID)
+
+		if data.ContactId.ValueInt64() != domain.RegistrantID {
+			diags.AddWarning(
+				fmt.Sprintf(`the contact_id does not match your local state and what you have at DNSimple for: domain=%s config_contact_id=%d remote_contact_id=%d.
```
Comment (wilson-dnsimple) [inline internal/framework/resources/registered_domain/common.go:129]:
```suggestion
				fmt.Sprintf(`The contact_id does not match your local state and what you have at DNSimple for: domain=%s config_contact_id=%d remote_contact_id=%d.
```

File: internal/framework/resources/registered_domain/common.go [hunk 4]
```diff
@@ -121,7 +123,17 @@ func (r *RegisteredDomainResource) updateModelFromAPIResponse(ctx context.Contex
...
+This can be due to a contact change that has not been reflected in your local config,
+if you have changed your local config due to a registered_domain_contact change.
+It may be that the contact change has not yet been completed.
+Until the change is compelted successfully you will continue to receive this warning.`, data.Name.ValueString(), data.ContactId.ValueInt64(), domain.RegistrantID),
```
Comment (wilson-dnsimple) [inline internal/framework/resources/registered_domain/common.go:133]:
```suggestion
Until the change is completed successfully you will continue to receive this warning.`, data.Name.ValueString(), data.ContactId.ValueInt64(), domain.RegistrantID),
```

File: internal/framework/resources/registered_domain_contact/common.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		registrantChangeResponse, err := r.config.Client.Registrar.GetRegistrantChange(ctx, r.config.AccountID, registrantChangeId)
+
+		if err != nil {
+			return err, false
+		}
+
+		if registrantChangeResponse.Data.State == consts.RegistrantChangeStateCancelled || registrantChangeResponse.Data.State == consts.RegistrantChangeStateCancelling {
+			diagnostics.AddError(
+				fmt.Sprintf("failed to change registrant for DNSimple Domain: %s, registrant change id: %d", data.DomainId.ValueString(), registrantChangeId),
+				"registrant change was cancelled, please investigate why this happened. You can refer to our support article https://support.dnsimple.com/articles/changing-domain-contact/ to get started and if you need assistance, please contact support at support@dnsimple.com",
```
Comment (wilson-dnsimple) [inline internal/framework/resources/registered_domain_contact/common.go:57]:
```suggestion
				"Registrant change was cancelled, please investigate why this happened. You can refer to our support article https://support.dnsimple.com/articles/changing-domain-contact/ to get started and if you need assistance, please contact support at support@dnsimple.com.",
```

File: internal/framework/resources/registered_domain_contact/common.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			)
+			return nil, true
+		}
+
+		if registrantChangeResponse.Data.State != consts.DomainStateRegistered {
+			tflog.Info(ctx, fmt.Sprintf("[RETRYING] Registrant change is not complete, current state: %s", registrantChangeResponse.Data.State))
+
+			return fmt.Errorf("registrant change is not complete, current state: %s. You can try to run terraform again to try and converge the registrant change", registrantChangeResponse.Data.State), false
```
Comment (wilson-dnsimple) [inline internal/framework/resources/registered_domain_contact/common.go:65]:
```suggestion
			return fmt.Errorf("Registrant change is not complete, current state: %s. You can try to run terraform again to try and converge the registrant change", registrantChangeResponse.Data.State), false
```

File: go.mod (hunk 1 above)
Comment (DXTimer) [reply go.mod]: Changed in [5e239dd](https://github.com/dnsimple/terraform-provider-dnsimple/pull/142/commits/5e239dda06b9d0bed9c1d101cbce2a7432446769)

**Review by wilson-dnsimple — APPROVED** [review]

**Pull request merged**